extensions/route_map/lib.rs

```diff
@@ -1,16 +1,358 @@
+            HandlerType::HttpDelete => "DELETE",
+            HandlerType::HttpPatch => "PATCH",
+            HandlerType::HttpPut => "PUT",
+            HandlerType::HttpHead => "HEAD",
+            HandlerType::HttpOther(s) => s,
+        };
+        f.write_str(s)
+    }
+}
+
+/// A node for the trie
+#[derive(Debug, Clone, Default)]
+struct Node {
+    // Map from path component to node
+    children: HashMap<String, Node>,
+    // Child for a placeholder
+    placeholder_child: Option<Box<Node>>,
+    handler_group: Option<HandlerGroup>,
+}
+
+/// A path router based on a prefix tree / trie.
+///
+/// Stores a handler references and other metadata for each node,
+/// as well as both static paths and plain routes for use with Starlite.
+///
+/// Routes can be added using `add_routes`.
+/// Given a scope containing a path, can retrieve handlers using `parse_scope_to_route`.
+#[pyclass]
+#[derive(Debug)]
+pub(crate) struct RouteMap {
+    ctx: wrappers::StarliteContext,
+    static_paths: HashSet<String>,
+    plain_routes: HashMap<String, HandlerGroup>,
+    root: Node,
+}
+
+// The functions below are available to Python code
+#[pymethods]
+impl RouteMap {
+    /// Creates an empty `RouteMap`
+    #[new]
+    #[args(debug = false)]
+    fn new(py: Python, debug: bool) -> PyResult<Self> {
+        Ok(RouteMap {
+            ctx: wrappers::StarliteContext::fetch(py, debug)?,
+            static_paths: HashSet::new(),
+            plain_routes: HashMap::new(),
+            root: Node::default(),
+        })
+    }
+
+    /// Adds a new static path by path name
+    fn add_static_path(&mut self, path: &str) {
+        self.static_paths.insert(path.to_string());
+    }
+
+    /// Checks if a given path refers to a static path
+    fn is_static_path(&self, path: &str) -> bool {
+        self.static_paths.contains(path)
+    }
+
+    /// Removes a path from the static path set
+    fn remove_static_path(&mut self, path: &str) -> bool {
+        self.static_paths.remove(path)
+    }
+
+    /// Add a collection of routes to the map
+    #[pyo3(text_signature = "($self, routes)")]
+    fn add_routes<'a>(&mut self, py: Python<'a>, routes: Vec<wrappers::Route>) -> PyResult<()> {
+        for route in routes {
+            self.add_route(py, route)?;
+        }
+        Ok(())
+    }
+
+    /// Add a route to the map
+    #[pyo3(text_signature = "($self, route)")]
+    fn add_route<'a>(&mut self, py: Python<'a>, route: wrappers::Route<'a>) -> PyResult<()> {
+        let path = route.path()?;
+        let path_parameters: &PyAny = route.path_parameters()?;
+        let is_static = self.static_paths.contains(path);
+
+        let new_handler_group = if is_static {
+            if !route.is_asgi(&self.ctx)? {
+                return Err(wrappers::ImproperlyConfiguredException::new_err(format!(
+                    "Static route handlers must be asgi handlers at {path}"
+                )));
+            }
+            if path_parameters.is_true()? {
```

Comment:
1. I wish this method was `PyAny::is_truthy()`.
2. Yeah. I was trying to be more general (`is_empty` requires a sequence/mapping, truthyness doesn't), but we already require it to be a sequence because we extract a vec from it. Will use `is_empty`.